contracts/JBGenericBuybackDelegate.sol

```diff
@@ -411,178 +378,154 @@ contract JBGenericBuybackDelegate is
-     *
-     * @param  _data the didPayData passed by the terminal
-     * @param  _minimumReceivedFromSwap the minimum amount received, to prevent slippage
-     */
-    function _swap(JBDidPayData3_1_1 calldata _data, uint256 _minimumReceivedFromSwap, IERC20 _projectToken)
-        internal
-        returns (uint256 _amountReceived)
-    {
-        address _terminalToken =
-            _data.forwardedAmount.token == JBTokens.ETH ? address(WETH) : _data.forwardedAmount.token;
-
-        bool _projectTokenIs0 = address(_projectToken) < _terminalToken;
-
+    /// @notice Swap the terminal token to receive the project token.
+    /// @param _data The didPayData passed by the terminal.
+    /// @param _minimumSwapAmountOut The minimum amount of project tokens received from the swap.
+    /// @param _amountToSwapWith The amount of tokens that are being used with which to make the swap.
+    /// @param _terminalToken The token paid in being used to swap.
+    /// @param _projectTokenIs0 A flag indicating if the pool will reference the project token as the first in the pair.
+    /// @return _amountReceived The amount of tokens received from the swap.
+    function _swap(
+        JBDidPayData3_1_1 calldata _data,
+        uint256 _minimumSwapAmountOut,
+        uint256 _amountToSwapWith,
+        address _terminalToken,
+        bool _projectTokenIs0
+    ) internal returns (uint256 _amountReceived) {
+        // Get a reference to the pool that'll be used to make the swap.
         IUniswapV3Pool _pool = poolOf[_data.projectId][_terminalToken];
 
-        // Pass the token and min amount to receive as extra data
+        // Try swapping.
         try _pool.swap({
             recipient: address(this),
             zeroForOne: !_projectTokenIs0,
-            amountSpecified: int256(_data.forwardedAmount.value),
+            amountSpecified: int256(_amountToSwapWith),
             sqrtPriceLimitX96: _projectTokenIs0 ? TickMath.MAX_SQRT_RATIO - 1 : TickMath.MIN_SQRT_RATIO + 1,
-            data: abi.encode(_data.projectId, _minimumReceivedFromSwap, _terminalToken, _projectToken)
+            data: abi.encode(_data.projectId, _terminalToken)
         }) returns (int256 amount0, int256 amount1) {
-            // Swap succeded, take note of the amount of PROJECT_TOKEN received (negative as it is an exact input)
+            // If the swap succeded, take note of the amount of tokens received. This will return as negative since it is an exact input.
             _amountReceived = uint256(-(_projectTokenIs0 ? amount0 : amount1));
         } catch {
-            // implies _amountReceived = 0 -> will later mint when back in didPay
-            return _amountReceived;
+            // If the swap failed, return.
+            return 0;
         }
 
-        // Burn the whole amount received
+        // Make sure the slippage is tolerable.
+        if (_amountReceived < _minimumSwapAmountOut) {
+            revert JuiceBuyback_MaximumSlippage();
```

Comment:
Aren't we reverting every max slippage with this one then (ie including twap-based ones) @mejango ?